core/src/main/java/org/apache/iceberg/V2Metadata.java

```diff
@@ -310,19 +310,24 @@ public Object get(int i) {
         case 1:
           return wrapped.snapshotId();
         case 2:
-          if (wrapped.sequenceNumber() == null) {
-            // if the entry's sequence number is null, then it will inherit the sequence number of
-            // the current commit.
+          if (wrapped.dataSequenceNumber() == null) {
+            // if the entry's data sequence number is null,
+            // then it will inherit the sequence number of the current commit.
             // to validate that this is correct, check that the snapshot id is either null (will
-            // also be inherited) or
-            // that it matches the id of the current commit.
+            // also be inherited) or that it matches the id of the current commit.
             Preconditions.checkState(
                 wrapped.snapshotId() == null || wrapped.snapshotId().equals(commitSnapshotId),
                 "Found unassigned sequence number for an entry from snapshot: %s",
                 wrapped.snapshotId());
+
+            // inheritance should work only for ADDED entries
+            Preconditions.checkState(
```

Comment:
This is new validation. Any cases when it is not safe to validate this? This class is used during writes.

Reply:
This sounds good to me.